qiskit/utils/parallel.py

```diff
@@ -50,101 +54,227 @@
 
 
-def get_platform_parallel_default():
-    """
-    Returns the default parallelism flag value for the current platform.
+CONFIG = user_config.get_config()
 
-    Returns:
-        parallel_default: The default parallelism flag value for the
-        current platform.
 
-    """
-    # Default False on Windows
-    if sys.platform == "win32":
-        parallel_default = False
-    # On macOS default false on Python >=3.8
-    elif sys.platform == "darwin":
-        parallel_default = False
-    # On linux (and other OSes) default to True
+def _task_wrapper(param):
+    (task, value, task_args, task_kwargs) = param
+    return task(value, *task_args, **task_kwargs)
+
+
+def _physical_cpus_assuming_twofold_smt():
+    if (sched_getaffinity := getattr(os, "sched_getaffinity", None)) is not None:
+        # It is callable, just pylint doesn't recognise it as `os.sched_getaffinity` because of the
+        # `getattr`.
+        # pylint: disable=not-callable
+        num_cpus = len(sched_getaffinity(0))
     else:
-        parallel_default = True
+        num_cpus = os.cpu_count() or 1
+    return (num_cpus // 2) or 1
 
-    return parallel_default
 
+def _parallel_default():
+    # We default to False on `spawn`-based multiprocessing implementations, True on everything else.
+    if (set_start_method := multiprocessing.get_start_method(allow_none=True)) is None:
+        # The method hasn't been explicitly set, but it would be badly behaved of us to set it for
+        # the user, so handle platform defaults.
+        return sys.platform not in ("darwin", "win32")
+    return set_start_method in ("fork", "forkserver")
 
-CONFIG = user_config.get_config()
 
-if os.getenv("QISKIT_PARALLEL", None) is not None:
-    PARALLEL_DEFAULT = os.getenv("QISKIT_PARALLEL", None).lower() == "true"
-else:
-    PARALLEL_DEFAULT = get_platform_parallel_default()
+@functools.cache
+def default_num_processes() -> int:
+    """Get the number of processes that a multiprocessing parallel call will use by default.
 
-# Set parallel flag
-if os.getenv("QISKIT_IN_PARALLEL") is None:
-    os.environ["QISKIT_IN_PARALLEL"] = "FALSE"
+    Such functions typically also accept a ``num_processes`` keyword argument that will supersede
+    the value returned from this function.
 
-if os.getenv("QISKIT_NUM_PROCS") is not None:
-    CPU_COUNT = int(os.getenv("QISKIT_NUM_PROCS"))
-else:
-    CPU_COUNT = CONFIG.get("num_process", local_hardware_info()["cpus"])
+    In order of priority (highest to lowest), the return value will be:
 
+    1. The ``QISKIT_NUM_PROCS`` environment variable, if set.
+    2. The ``num_processes`` key of the Qiskit user configuration file, if set.
+    3. Half of the logical CPUs available to this process, if this can be determined.  This is a
+       proxy for the number of physical CPUs, assuming two-fold simultaneous multithreading (SMT);
+       emperically, multiprocessing performance of Qiskit seems to be worse when attempting to use
```

Comment:
```suggestion
       empirically, multiprocessing performance of Qiskit seems to be worse when attempting to use
```

Unless this is a British spelling I'm not aware of 😄.